### PR TITLE
Added OpenTelemetry support

### LINKED ISSUE #5052: Otel activation for message traces
### Proposed change

In https://github.com/nats-io/nats-server/pull/5014 a basic capability to do on-demand message tracing was introduced for core pub-sub messages.

This capability is activated using a NATS specific header.  The bigger picture is that users want to trigger traces based on trace headers from the likes of Open Telemetry etc.

There is prior-art in the server around [ADR-3](https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-3.md) that already supports triggering based on the headers we need and there are already functions for parsing the headers and determining if it's a continuation of a trace that started outside of NATS. See the `shouldSample()` function.

This leaves the question of where should traces be delivered when these headers like `traceparent` are present, of course we don't know as previously we configured the delivery per message. The only option I can think of is to add an account scope config item that configures a per-account subject for these externally-activated traces to be delivered.  

Some other cases to consider, there might be more:

 * If the configuration is not set the account does not support this style of activation of tracing.
 * If the `Nats-Trace-Only` is present AND otel headers are set, we honor `Nats-Trace-Only` and do not deliver the message
 * If `Nats-Trace-Dest` is set AND otel headers are set, we honor `Nats-Trace-Dest` and deliver to that subject

/cc @kozlovic 

### Use case

Integration with external tracing systems on account level.

### Contribution

_No response_

### PR DESCRIPTION
**Pull request by kozlovic**
If the `Nats-Trace-Dest` header is not present, but `Traceparent` is and its last token is `01`, then message tracing is triggered. This also requires that the account be defined with a `trace_dest` subject so that traces can be sent there.
Note that `Nats-Trace-Only` is not applicable for `Traceparent`.

Addition to PR #5014
Resolves #5052

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>

### CODE REVIEW

File: server/server.go
Comment (derekcollison) [inline server/server.go:1054]: `IsValidPublishSubject`?

File: server/msgtrace.go
Comment (derekcollison) [inline server/msgtrace.go:579]: Can inline I think.

```go
for i, pos := range [2]string{MsgTraceSendTo, trcCtx}
```

File: server/opts.go
Comment (derekcollison) [inline server/opts.go:3030]: Same here?

`IsValidPublishSubject`

**Review by derekcollison — COMMENTED** [review]
In general LGTM - A few nits/questions.

File: server/msgtrace.go
Comment (kozlovic) [reply server/msgtrace.go:579]: No, we are given `positions` which are the indexes in the `msg` byte slice where to put back the first character of the 2 given headers. But agreed that I can replace `headers` with a slice of the first character of the 2 headers.

File: server/server.go
Comment (kozlovic) [reply server/server.go:1054]: That is a good point and will add to the test that I will reject wildcard subjects (which I assume `IsValidPublishSubject` does.

**Comment by kozlovic** [comment]
@derekcollison Updated the PR based on your comments. But also would want @ripienaar to have a look at it tomorrow to see if that matches its expectations for otel support.

**Comment by derekcollison** [comment]
Thanks @kozlovic, agree will wait on @ripienaar.

File: server/msgtrace.go
Comment (ripienaar) [inline server/msgtrace.go:348]: I'll do some testing but just fyi, looks like the initial implemenation in the server was maybe a bit naive and I based my spec on that.

This isnt just a case of a `01` and if its `01` its traced, this is actually a hex encoded bit field - today there is only one value stored in this bit field so the naive check is ok, but we might need to revisit this in future.  And I'll try to get some example trace headers

**Review by ripienaar — CHANGES_REQUESTED** [review]
This looks good to me but keeping in mind that one comment.

Bigger problem is that I dont think our otel support ever worked, when we wrote it not much was using the then-new standard :)

https://github.com/nats-io/nats-server/blob/b2d37d7a7d5dda2b54fba70763c6f4aeeaea9b6e/server/accounts.go#L2067

This will do `Traceparent` and all our checks are done case sensitively, but the header is exactly `traceparent` in the spec, we need to either check without case sensitivity as the http spec requires or at least make this one lowercase in `trcCtx`.

I've asked for example headers so will update when I got those to be 100% certain

So hand fixing that it works with lowercase `traceparent` but I see in the headers we send on its again upper cased

From a trace:
```
{"header":{"Nats-Trace-Hop":["2"],"Traceparent":.....
```

Not sure if this is avoidable because go headers class is not great but ideally we change the case of headers

**Comment by kozlovic** [comment]
@ripienaar The original code with "Traceparent" is correct because the checks were done after "lifting" the message header into a http.Header that makes it canonical. 
The issue that I have is that I use low level "getHeader" to just find the string without having to lift the headers. That will be a problem since bytes.IndexOf() does not do case insensitive. Of course, in this case I could have a version of the header string that is all lower case, but if the app sends with a Go library, it will likely be canonical. So that is a huge problem, and maybe we unfortunately have to pay the price of lifting the headers?

File: server/msgtrace.go
Comment (kozlovic) [reply server/msgtrace.go:348]: Ok, thanks. I'll check the spec too and see what needs to be done there.

**Comment by ripienaar** [comment]
> @ripienaar The original code with "Traceparent" is correct because the checks were done after "lifting" the message header into a http.Header that makes it canonical. The issue that I have is that I use low level "getHeader" to just find the string without having to lift the headers. That will be a problem since bytes.IndexOf() does not do case insensitive. Of course, in this case I could have a version of the header string that is all lower case, but if the app sends with a Go library, it will likely be canonical. So that is a huge problem, and maybe we unfortunately have to pay the price of lifting the headers?

Would it be cheaper to do `getHeader` on both the `traceparent` and then `Traceparent`? Probably unavoidable that'd we'd send it on as `Traceparent` though, I wonder if the std otel libraries will handle that well let me try and see

**Comment by ripienaar** [comment]
> Would it be cheaper to do `getHeader` on both the `traceparent` and then `Traceparent`? Probably unavoidable that'd we'd send it on as `Traceparent` though, I wonder if the std otel libraries will handle that well let me try and see

Looks to me like we'd have to send it as `traceparent` in the headers contained in the traces, but probably we cant because of go:

https://github.com/nats-io/nats-server/blob/b2d37d7a7d5dda2b54fba70763c6f4aeeaea9b6e/server/msgtrace.go#L54

We might have to document that people will have to lowercase it to use it :(

**Comment by ripienaar** [comment]
Apparently if one use the header like a map you can circumvent the uppercase nonsense https://go.dev/play/p/WnjqoLSLhe

**Comment by kozlovic** [comment]
@ripienaar I made the requested changes, please check to see if that works for you.

**Review by ripienaar — APPROVED** [review]
LGTM, does what I think we need from a behaviour perspective. But lets hold off on the merge while we get some details from one user to confirm it's 100% the right track

**Comment by ripienaar** [comment]
The only possible addition I could think of is to be able to force compression on those messages - but from what I saw compression isnt helping much on these anyway so no biggie.

**Comment by kozlovic** [comment]
@ripienaar Your opinion on changing the constant name `MsgTraceSendTo` (which maps to `Nats-Trace-Dest` to `MsgTraceDest`? Since this PR already "breaks" application using main in that the RequestInfo.Header is no longer an http.Header, may as well take that opportunity?

PS: I made them public thinking of apps importing server code to do tracing, let me know if you think it was a bad idea.

**Comment by ripienaar** [comment]
> @ripienaar Your opinion on changing the constant name `MsgTraceSendTo` (which maps to `Nats-Trace-Dest` to `MsgTraceDest`? Since this PR already "breaks" application using main in that the RequestInfo.Header is no longer an http.Header, may as well take that opportunity?
> 
> PS: I made them public thinking of apps importing server code to do tracing, let me know if you think it was a bad idea.

Yes go ahead, no1 is using this other than me - its quite hard to use without docs which I'll add later hehe

Feel free to change whatever you need.  Public is fine I think

**Comment by ripienaar** [comment]
Functionality wise I am happy to merge this we have confirmation this will do what we intend.

So will wait for code review from @derekcollison

**Review by derekcollison — APPROVED** [review]
LGTM

**Comment by kozlovic** [comment]
@derekcollison @ripienaar I would like to make sure of something before this gets into a future release and cannot be changed. When I added the account's trace destination, I made it a public field because the way options are defined. Meaning that if one want to programmatically have an account and configure it, one would do: 
```
acc := NewAccount("MyAccount")
acc.TraceDest = "my.trace.dest"
opts.Accounts = []*Account{acc}
...
s, err := NewServer(opts)
...
```
Hence my decision to make `TraceDest` public, but is it the right thing? Are users expected to be able to configure accounts programmatically? It does not seem to be many public fields in `Account`, so I am now wondering if I made a mistake? 

Again, it would not be too late to change since this has not been released yet.

Conversely, I think that TraceDest *needs* to be added to `Account` in the JWT library in order for JWT account updates can set/update this value (this is what I am working on right now, along with the AllowTrace in service/stream).

**Comment by ripienaar** [comment]
Where I embed NATS I often wish I had more options to configure accounts programatically - very little one can do now and it can be awkward to do things with accounts.

That said, I dont know if thats a wider need or not, certainly people who embed the server isnt a huge demographic so I'll take it or leave it, no strong oppinions

**Comment by derekcollison** [comment]
Can it be access without a lock? If not then would make private and maybe add setter and getter.

**Comment by kozlovic** [comment]
> Can it be accessed without a lock? If not then would make private and maybe add setter and getter.

That's a good point. I will switch to private.

**Pull request merged**
